fix: issue where provider engine needs to rerun after permissions are granted to ensure correct proxies are assigned

adonesky1 opened the pull request:
## **Description**
When the "Selected networks for each site" toggle is enabled, the wallet API should use an origin-specific chain for all interactions. However this doesn't work correctly if the connection status has changed since the last refresh. If you grant permissions to a site or revoke them (i.e. connect to or disconnect from a site), the site does not work as expected until after the site is refreshed.

**Expected behavior**
The wallet API should always use the dapp-selected chain for sites with permissions, and should use the globally-selected chain for sites without permissions. This should apply even if the permissions have changed since the page loaded.

**Solution**

This PR introduces [changes made to `@metamask/selected-network-controller`](https://github.com/MetaMask/core/blob/main/packages/selected-network-controller/CHANGELOG.md#changed) to resolve the issue described above:
> Previously the SelectedNetworkController only constructed proxies for domains that had permissions. Other domains have no associated proxy and the getProviderAndBlockTracker method would throw an error. This was problematic because we grab the network client for an origin a single time when constructing an RPC pipeline for that origin in the MetaMask extension. We don't re-create the RPC pipeline when permissions change. That means that the pipeline is setup with the wrong network client and cannot be updated. The following changes ensure seamlessly proxying calls during sessions where a dapp connects/disconnects and provides a path for clients to prune inactive proxies.

This domainProxyMap param now passed into `SelectedNetworkController` allows the client to handle pruning inactive proxies while the SelectedNetworkController handles adding entries - it can't handle removal, as it doesn't know which connections are active.

The extension passes in an instance of a newly created `WeakStringMap` class which allows for garbage collection of its data once no longer referenced.

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/23509

## **Manual testing steps**

The problem presented itself in two scenarios: when granting site permissions, and when revoking all site permissions.

**When granting permissions:**
- Enable the "Selected networks for each site" toggle
- Open a site with no permissions in one tab, and the fullscreen wallet UI in another
- If the site was open already, hard-refresh before proceeding
- Call eth_requestAccounts on the site, asking for the eth_accounts permissions
- Approve the request
- Call eth_chainId from the site, and see that it matches the wallet selected chain
- Switch the selected chain in the fullscreen UI
- Call eth_chainId from the site, and see that it is still on the same chain the wallet was connected to before switching

**When revoking permissions:**
- Redo all the same steps for **When granting permissions:**
- Open the browser-action popup, click on the three-vertical-dots menu, click "Connected sites", then disconnect from the site
- Call eth_chainId from the site. See that the sites connected chain is now the same as the "globally selected chain" in the wallet

## **Screenshots/Recordings**


### **Before**

This recording shows both reproduction use cases (granting permissions, then revoking)

[queue-selected-chain-bug-2024-03-14_19.49.19.webm](https://github.com/MetaMask/metamask-extension/assets/2459287/092c0a9d-e734-4309-9455-d6444f7504ef)

### **After**

https://github.com/MetaMask/metamask-extension/assets/34557516/62145d32-493b-41a2-ba8e-a1545ef1286a


## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x]I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

adonesky1 committed:
Fix issue where provider engine needs to rerun after permissions are granted to ensure correct proxies are assigned

adonesky1 committed:
add WeakStringMap class for domainProxyMap on SelectedNetworkController

adonesky1 commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

adonesky1 committed:
lint + dedupe

adonesky1 commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

adonesky1 committed:
WeakStringMap is broken

adonesky1 committed:
remove finalizationRegistry

adonesky1 committed:
update weakstringmap

adonesky1 committed:
Wrap properties of WeakStringMap values in WeakRefs instead of the value itself

adonesky1 committed:
fix types

adonesky1 commented:
@metamaskbot update-policies

adonesky1 committed:
update patch

adonesky1 committed:
update yarn.lock

adonesky1 commented:
@metamaskbot update-policies

jiexi commented on .yarn/patches/@metamask-network-controller-npm-18.1.0-680908c29a.patch:
```diff
@@ -1,23 +1,23 @@
-diff --git a/dist/chunk-VNCJZRDU.js b/dist/chunk-VNCJZRDU.js
```
Making sure this patch change was intended. I'm not actually clear on how these patches are generated and maintained

adonesky1 replied:
Yes this was intended. Its updated to point at the latest networkController version which we want to bump to here. It is required to pass [this recently added test](https://github.com/MetaMask/metamask-extension/blob/eb93c9847cbb0769a260d53e70262f32a51e9cb9/test/e2e/tests/onboarding/onboarding.spec.js#L311) (for which the patch was added)

adonesky1 committed:
add tests

jiexi commented on app/scripts/lib/WeakDomainProxyMap.ts:
```diff
@@ -0,0 +1,58 @@
+type WeakReffedValueObj = {
```
Do you think it's worth making this file more generic?

```
type WeakRefObject<T> = {
  [K in keyof T]: WeakRef<T[K]>;                            
};

...

export class WeakRefObjectMap<T> {
  map: Map<string, WeakRefObject<T>>;
  
...
```

adonesky1 replied:
Yeah we sure could! I'll go ahead and do that!

adonesky1 committed:
WeakDomainProxyMap -> WeakRefObjectMap

jiexi commented:
Tested working

jiexi commented on app/scripts/lib/WeakRefObjectMap.ts:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  get(key: string) {
+    const weakReffedValue = this.map.get(key) as
+      | WeakRefCompatibleObject<Record<string, object>>
+      | undefined;
```
is the as needed here?

adonesky1 replied:
No sir! Removed!

adonesky1 committed:
remove unnecessary cast

adonesky1 committed:
add e2e test

adonesky1 committed:
fix e2e test

adonesky1 committed:
replace e2e test

adonesky1 committed:
cleanup

adonesky1 commented on test/e2e/tests/request-queuing/chainid-check.spec.js:
```diff
@@ -0,0 +1,132 @@
+const { PAGES } = require('../../webdriver/driver');
+
+describe('Request-queue chainId proxy sync', function () {
+  it('should preserve per dapp network selections after connecting without refresh calls @no-mmi', async function () {
```
this test fails against develop (you can pull down and test it on [this branch](https://github.com/MetaMask/metamask-extension/compare/chainid-check?expand=1)) because the proxy doesn't update correctly without a refresh prior to the changes in this PR

adonesky1 committed:
add forEach to WeakRefObjectMap

adonesky1 committed:
cleanup rebase

adonesky1 commented:
@metamaskbot update-policies

adonesky1 commented on app/scripts/lib/WeakRefObjectMap.ts:
```diff
@@ -0,0 +1,66 @@
+    this.map.clear();
+  }
+
+  forEach(callback: (value: WeakRefCompatibleObject<Record<string, object>>, key: string, map: Map<string, WeakRefCompatibleObject<Record<string, object>>>) => void, thisArg?: any): void {
```
Adding this pass through for forEach to access the internal map for this custom Map wrapper [so that it can be accessed cleanly here](https://github.com/MetaMask/core/pull/4130/files#r1558425789) and preserve the Map typing for the `#domainProxyMap` variable on the `SelectedNetworkController` where its used...

adonesky1 committed:
lint

jiexi commented on app/scripts/lib/WeakRefObjectMap.ts:
```diff
@@ -0,0 +1,74 @@
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    thisArg?: any,
+  ): void {
+    this.map.forEach((value, key) => {
+      if (thisArg) {
+        callback.call(thisArg, value, key, this.map);
+      } else {
+        callback(value, key, this.map);
+      }
+    });
```
should this be calling `this.get()` and skipping undefineds?

adonesky1 replied:
> should this be calling this.get()
Probably...? Though you'll notice that in practice we're only referencing the key/domain where we call `forEach`: https://github.com/MetaMask/core/pull/4130/files#diff-5e705302d595033f24cac5f20d14ecf307709fd61c46a6d01bf9519bf5b91d86R217

adonesky1 replied:
But I can make this update

adonesky1 replied:
Actually I don't really think it makes sense to get the dereffed instance, especially since it could end up mutating the collection if it finds that a weak ref has been garbage collected. I think it could lead to unpredictable behavior, and its easy enough to go in and deref in callback if you need to.

jiexi replied:
To clarify, currently this loop over `this.map` will call `callback` for all values in the map, including values that have been garbage collected. Additionally, `callback` will be provided the WeakRef'ed object. My assumption would be that all WeakRef wrapping and unwrapping would be hidden from the consumer of `WeakRefObjectMap`. To the consumer, this class would work like a regular Map, except that entries could disappear

jiexi replied:
I do see that we only use the key/domain, but even in that case, should we be iterating over entries where the value has been garbage collected already?

jiexi commented on test/e2e/tests/request-queuing/chainid-check.spec.js:
```diff
@@ -0,0 +1,132 @@
+          `return window.ethereum.request(${chainIdRequest})`,
+        );
+
+        assert.equal(firstChainIdCallAfterConnect, '0x1'); // 1337
```
```suggestion
        assert.equal(firstChainIdCallAfterConnect, '0x1');
```

jiexi commented on test/e2e/tests/request-queuing/chainid-check.spec.js:
```diff
@@ -0,0 +1,132 @@
+          `return window.ethereum.request(${chainIdRequest})`,
+        );
+        // after connecting the dapp should now have its own selected network
+        // independent from the globally selected and therefore should have changed
```
```suggestion
        // independent from the globally selected and therefore should not have changed
```

should not have changed, right?

adonesky1 committed:
Update test/e2e/tests/request-queuing/chainid-check.spec.js

Co-authored-by: jiexi <jiexiluan@gmail.com>

adonesky1 committed:
Update test/e2e/tests/request-queuing/chainid-check.spec.js

Co-authored-by: jiexi <jiexiluan@gmail.com>

Gudahtt commented on app/scripts/lib/WeakRefObjectMap.ts:
```diff
@@ -0,0 +1,74 @@
+  }
+
+  set(key: string, value: Record<string, object>) {
+    const weakReffedValueObj = {} as WeakRefCompatibleObject<
+      Record<string, object>
+    >;
```
Nit: Here it would be safer to use a type annotation rather than a type assertion. Type assertions are an _override_, they provide no feedback if the override stops making sense or is no longer required. We're not getting as much value from the type system if we use an assertion. With an annotation. TypeScript will still type check it against the value and let us know if it's invalid.

See these draft TypeScript guidelines for details on this recommendation: https://github.com/MetaMask/contributor-docs/blob/85e06a3097ffdacb7b2fbda9d074955582660181/docs/typescript.md#provide-a-type-annotation--when-instantiating-an-empty-composite-data-type-value


```suggestion
    const weakReffedValueObj: WeakRefCompatibleObject<
      Record<string, object>
    > = {};
```

adonesky1 replied:
done here: https://github.com/MetaMask/metamask-extension/pull/23525/commits/94c4976528bb0f83c9815acd22be4a45fa4f5646

Gudahtt commented on app/scripts/lib/WeakRefObjectMap.ts:
```diff
@@ -0,0 +1,74 @@
+      return undefined;
+    }
+
+    const deReffedValue = {} as { [dereffedValueObjKey: string]: WeakKey };
```
Nit: A type annotation would be preferred here as well, for the same reason given in the other comment:

```suggestion
    const deReffedValue: { [dereffedValueObjKey: string]: WeakKey } = {};
```

adonesky1 replied:
done here: https://github.com/MetaMask/metamask-extension/pull/23525/commits/94c4976528bb0f83c9815acd22be4a45fa4f5646

Gudahtt commented on app/scripts/lib/WeakRefObjectMap.ts:
```diff
@@ -0,0 +1,74 @@
+  [P in keyof T]: T[P] extends object ? WeakRef<T[P]> : never;
+};
+
+export class WeakRefObjectMap {
```
Nit: It would be great to have TSDoc descriptions for this class and all public methods.

I get that it's mostly a Map implementation, but the goals of using WeakRef here are worth explaining. The part where we deconstruct the network client to store each property independently is super weird as well. I almost asked why we were doing that, before remembering the fact that we only use hard references to the network client properties not the client object itself.

adonesky1 replied:
done here: https://github.com/MetaMask/metamask-extension/pull/23525/commits/a3770013832099fc91e51e849f3ff399b556d4f8

Gudahtt commented on app/scripts/lib/WeakRefObjectMap.ts:
```diff
@@ -0,0 +1,74 @@
+      return undefined;
+    }
+
+    const deReffedValue = {} as { [dereffedValueObjKey: string]: WeakKey };
```
Nit: The capitalization here seems a bit inconsistent. With the name `dereffedValueObjKey` we're treating 'dereffed" as one word, but in `deReffedValue` we're treating "de" and "Reffed" as two separate words.

I think this should be `dereffedValue`, lower-casing the R.

adonesky1 replied:
done here: https://github.com/MetaMask/metamask-extension/pull/23525/commits/94c4976528bb0f83c9815acd22be4a45fa4f5646

Gudahtt commented on app/scripts/lib/WeakRefObjectMap.ts:
```diff
@@ -0,0 +1,74 @@
+    this.map.set(key, weakReffedValueObj);
+  }
+
+  get(key: string) {
```
Nit: I'd recommend providing explicit return types for any of the methods that return a value. This aids readability and ensures the return type doesn't change accidentally (this can be easy to do accidentally if we're relying on type inference)

See here for more on that recommendation: https://github.com/MetaMask/contributor-docs/blob/85e06a3097ffdacb7b2fbda9d074955582660181/docs/typescript.md#for-functions-and-methods-provide-explicit-return-types

adonesky1 replied:
done [here](https://github.com/MetaMask/metamask-extension/pull/23525/commits/a3770013832099fc91e51e849f3ff399b556d4f8) and [here](https://github.com/MetaMask/metamask-extension/pull/23525/commits/94c4976528bb0f83c9815acd22be4a45fa4f5646)

Gudahtt commented on app/scripts/lib/WeakRefObjectMap.ts:
```diff
@@ -0,0 +1,74 @@
+      key: string,
+      map: Map<string, WeakRefCompatibleObject<Record<string, object>>>,
+    ) => void,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Nit: Typically we encourage adding a comment each time we use `any` for some unavoidable reason, as we are trying to discourage its use. A comment can help explain why it's necessary so someone else doesn't waste time trying to remove it without understanding why.

In this case, I think `any` makes sense because this is an unbound method, so the `this` value is unknown, it can be anything. Also the `Map` type this is based on uses `any` for this parameter as well.

adonesky1 replied:
yep makes sense! Added a comment here: https://github.com/MetaMask/metamask-extension/pull/23525/commits/a3770013832099fc91e51e849f3ff399b556d4f8#diff-4d13f1f0ba4d949f6dfe026c7c50a982dfd0e3b43d3bbd265f6acdca8f501302R180

Gudahtt reviewed (commented):
Looks great!

The map doesn't yet have a type that is compatible with what the SelectedNetworkController needs from the `domainProxyMap`, so that seems worth fixing first. But I think that should be easy to resolve, and all the other comments I left were minor issues as well.

Gudahtt commented on app/scripts/lib/WeakRefObjectMap.ts:
```diff
@@ -0,0 +1,74 @@
+  [P in keyof T]: T[P] extends object ? WeakRef<T[P]> : never;
+};
+
+export class WeakRefObjectMap {
```
Actually even that wouldn't be totally type compatible, because the `SelectedNetworkController` needs a `NetworkProxy` for the type. But this class doesn't need to know anything about that. So this would be a great use case for a generic type parameter.

e.g. something like:

```
class WeakRefObjectMap<WeakReffedObject extends object> implements Map<string, WeakReffedObject> {
```

adonesky1 replied:
🤔 won't object suffice in that case? The object is not actually weakreffed its properties are?

adonesky1 replied:
done here: https://github.com/MetaMask/metamask-extension/pull/23525/commits/94c4976528bb0f83c9815acd22be4a45fa4f5646

Gudahtt replied:
It'll need to match the type here: https://github.com/MetaMask/core/blob/e86b84e6260beb689d95a4bed724d5a55237f35c/packages/selected-network-controller/src/SelectedNetworkController.ts#L101

That's why `object` wouldn't be specific enough; the controller expects it to be a map of `NetworkProxy` objects specifically

adonesky1 replied:
ah I see ok!

adonesky1 committed:
WeakRefObjectMap implements Map<string, object>

adonesky1 committed:
add TSdocs for WeakRefObjectMap

adonesky1 committed:
lint

Gudahtt commented on app/scripts/lib/WeakRefObjectMap.ts:
```diff
@@ -0,0 +1,193 @@
+
+  /**
+   * Executes a provided function once for each key-value pair in the map, in insertion order.
+   * Note that the values passed to the callback function are the `WeakRefCompatibleObject`s,
```
Hmm; making the type of the values in this callback would conflict with the types for `Map`.

It sounds potentially useful, but for type compatibility we might want to make this a separate method if it's something we need.

adonesky1 committed:
type jail

jiexi committed:
Update WeakRefObjectMap types (#24006)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24006?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: Alex Donesky <adonesky@gmail.com>

adonesky1 committed:
fix log in WeakRefObjectMap

adonesky1 committed:
fix tests

adonesky1 commented on test/e2e/tests/request-queuing/multiple-networks-dapps-txs.spec.js:
```diff
@@ -109,15 +109,13 @@ describe('Request Queuing for Multiple Dapps and Txs on different networks.', fu
 
         // Dapp one send tx
         await driver.switchToWindowWithUrl(DAPP_URL);
-        await driver.executeScript(`window.location.reload()`);
```
its no longer required to reload the page after connecting to ensure that the proxies are correctly synced

adonesky1 committed:
add iterator tests

jiexi commented on app/scripts/lib/WeakRefObjectMap.test.ts:
```diff
@@ -0,0 +1,146 @@
+      map.set('key1', { objKey1: { detail: 'value1' } });
+      map.set('key2', { objKey2: { detail: 'value2' } });
+    });
+    it('iterates over entries correctly', () => {
```
```suggestion

    it('iterates over entries correctly', () => {
```

jiexi committed:
Update app/scripts/lib/WeakRefObjectMap.test.ts

adonesky1 committed:
Merge branch 'develop' into ad/fix-selected-network-controller-proxies-bug

Gudahtt commented on app/scripts/lib/WeakRefObjectMap.ts:
```diff
@@ -0,0 +1,214 @@
+ * referenced so that they can be garbage collected if/when a dapp connection ends without effective cleanup.
+ */
+
+export class WeakRefObjectMap<T extends Record<string, object>>
```
Nit: I realize it is a common convention to use a single capital letter for type parameters, but we have been trying to avoid single-letter params in general lately, including for these. Our shared ESLint config disallows this in later versions (see https://github.com/MetaMask/eslint-config/pull/292)

adonesky1 replied:
Changed it here: https://github.com/MetaMask/metamask-extension/pull/23525/commits/9592db0d6c35dc613a7bab5bd7408b0737ac5781

to `RecordType`... did you have a particular name you'd prefer?

Gudahtt replied:
Seems alright to me!

Gudahtt commented on app/scripts/lib/WeakRefObjectMap.ts:
```diff
@@ -0,0 +1,214 @@
+   * @returns The `WeakRefObjectMap` instance.
+   */
+  set(key: string, value: T): this {
+    const weakRefValue: WeakRefObject<T> = {} as WeakRefObject<T>;
+    for (const keyValue in value) {
+      if (!Object.hasOwn(value, keyValue)) {
+        continue;
+      }
+      const item: T[typeof keyValue] = value[keyValue];
+      if (typeof item === 'object' && item !== null) {
+        weakRefValue[keyValue] = new WeakRef(item);
+      } else {
+        console.warn(
+          `Property ${String(
+            keyValue,
+          )} is not an object and cannot be weakly referenced.`,
+        );
+      }
+    }
+    this.map.set(key, weakRefValue);
```
I had an earlier comment about removing this type assertion in favour of an annotation. I see that that suggestion was merged, but the assertion was added back, so now we have both 😅.

I had misunderstood the need for the assertion in my original suggestion, I realize now why it was added. Using an annotation failed because TypeScript correctly identified that it was invalid, as the type is empty here. Really this is a _partial_ `WeakRefObject<T>`. It's not complete until after the loop.

This is similar to cases where we use `reduce` to build something up into a complete object; it's partial inside the `reduce` callback, but complete at the end. Yet TypeScript can't tell that it's complete, so we need to assert that it is.

The approach I usually use in cases like this is to declare the type as Partial to begin with, so that we have proper type checking inside the loop itself. Then at the point where we know it's complete, we can use an assertion to align the type with that. e.g.

```suggestion
    const weakRefValue: Partial<WeakRefObject<T>> = {};
    for (const keyValue in value) {
      if (!Object.hasOwn(value, keyValue)) {
        continue;
      }
      const item: T[typeof keyValue] = value[keyValue];
      if (typeof item === 'object' && item !== null) {
        weakRefValue[keyValue] = new WeakRef(item);
      } else {
        console.warn(
          `Property ${String(
            keyValue,
          )} is not an object and cannot be weakly referenced.`,
        );
      }
    }
    this.map.set(key, weakRefValue as WeakRefObject<T>);
```

adonesky1 replied:
Done here: https://github.com/MetaMask/metamask-extension/pull/23525/commits/9592db0d6c35dc613a7bab5bd7408b0737ac5781

Gudahtt commented on app/scripts/lib/WeakRefObjectMap.ts:
```diff
@@ -0,0 +1,214 @@
+      if (typeof item === 'object' && item !== null) {
+        weakRefValue[keyValue] = new WeakRef(item);
+      } else {
+        console.warn(
```
Should we throw an error here instead? Otherwise our type assertion would be lying, leaving missing properties in the type that we set.

adonesky1 replied:
Done here: https://github.com/MetaMask/metamask-extension/pull/23525/commits/9592db0d6c35dc613a7bab5bd7408b0737ac5781

adonesky1 committed:
address feedback: update types, throw error instead of console warning

Gudahtt commented on app/scripts/lib/WeakRefObjectMap.ts:
```diff
@@ -0,0 +1,218 @@
+      return undefined;
+    }
+
+    const deRefValue: RecordType = {} as RecordType;
+    for (const keyValue in weakRefValue) {
+      if (!Object.hasOwn(weakRefValue, keyValue)) {
+        continue;
+      }
+      const deref = weakRefValue[keyValue].deref();
+      if (deref === undefined) {
+        this.map.delete(key);
+        return undefined;
+      }
+      deRefValue[keyValue] = deref;
+    }
+
+    return deRefValue;
```
Nit: One more case where we can improve type safety by moving the assertion to after we've finished assembling the object, using `Partial` in the interim
```suggestion
    const deRefValue: Partial<RecordType> = {};
    for (const keyValue in weakRefValue) {
      if (!Object.hasOwn(weakRefValue, keyValue)) {
        continue;
      }
      const deref = weakRefValue[keyValue].deref();
      if (deref === undefined) {
        this.map.delete(key);
        return undefined;
      }
      deRefValue[keyValue] = deref;
    }

    return deRefValue as RecordType;
```

Gudahtt reviewed (dismissed):
LGTM! One more pending suggestion (re: improving types), but overall this looks great

adonesky1 committed:
Update app/scripts/lib/WeakRefObjectMap.ts

Co-authored-by: Mark Stacey <markjstacey@gmail.com>
